quinn/src/endpoint.rs

```diff
@@ -384,20 +384,20 @@ pub(crate) struct Shared {
 impl State {
     fn drive_recv<'a>(&'a mut self, cx: &mut Context, now: Instant) -> Result<bool, io::Error> {
         self.recv_limiter.start_cycle();
+        let recv_count = recv_datagram_count(&self.udp_state) / self.udp_state.gro_segments();
         let mut metas = [RecvMeta::default(); BATCH_SIZE];
-        let mut iovs = MaybeUninit::<[IoSliceMut<'a>; BATCH_SIZE]>::uninit();
+        let mut iovs = std::array::from_fn::<MaybeUninit<IoSliceMut<'a>>, BATCH_SIZE, _>(|_| {
+            MaybeUninit::uninit()
+        });
         self.recv_buf
-            .chunks_mut(self.recv_buf.len() / BATCH_SIZE)
+            .chunks_mut(self.recv_buf.len() / recv_count)
             .enumerate()
-            .for_each(|(i, buf)| unsafe {
-                iovs.as_mut_ptr()
-                    .cast::<IoSliceMut>()
-                    .add(i)
-                    .write(IoSliceMut::<'a>::new(buf));
+            .for_each(|(i, buf)| {
+                iovs[i].write(IoSliceMut::<'a>::new(buf));
             });
-        let mut iovs = unsafe { iovs.assume_init() };
+        let iovs = unsafe { slice_assume_init_mut(&mut iovs[..recv_count]) };
```

Comment:
If gro is available, then recv_count is 1. But AIUI, in a single iovec, you can only receive messages from the same source. So this means that in each syscall you can receive messages from at most one remote.

Would it make more sense to split this more dynamically? E.g. if we detect burst activity, create 1 big iovec with 64 gro segments. But if activity is more spread out, use many iovecs with one gro segment each.

Reply:
Good catch; this is probably a regression for endpoints with many active connections. However, per https://github.com/quinn-rs/quinn/pull/1354, if we're using GRO we must accept exactly 64 segments, or else we'll get inadvertent packet truncation. I guess we'll have to bite the bullet and lower the max datagram size...

Reply:
Why is the current default 64k? Is this useful outside of communication on localhost?

Reply:
The default is [from RFC 9000](https://www.rfc-editor.org/rfc/rfc9000.html#transport-parameter-definitions), though we're free to pick something else. Whatever value we choose will impose a hard upper limit to MTU detection, and it'd be nice to Just Work on networks with jumbo frames. However, AFAIK such networks are rare, so I'm willing to drop this to a best-case internet MTU and let jumbo frame users adjust their configurations if necessary.

Reply:
Do we have to allocate per segment or can we allocate a large buffer which would allow some jumbo frames provided that most frames are smaller?

Reply:
If we receive a GRO batch consisting of many jumbo frame packets, and we did not allocate enough space for all of them, packets may be truncated (i.e. lost) per https://github.com/quinn-rs/quinn/pull/1354.

Reply:
Consensus in the implementers' slack seems to be that `recvmmsg` isn't adding a lot here, i.e. we probably won't see much regression from doing one syscall per sender as in this PR. This is in principle workload dependent... it's tempting to leverage that to simplify our APIs a bunch, though.

Reply:
Yeah, sounds like ditching `recvmmsg()` could make sense.